File: apps/backend-java/src/main/java/com/housingplatform/admin/AdminService.java
package com.housingplatform.admin;

import com.housingplatform.persistence.repository.AdminRepositoryCustom;
import com.housingplatform.admin.dto.AdminDashboardStats;
import com.housingplatform.admin.dto.AdminHost;
import com.housingplatform.admin.dto.AdminPayment;
import com.housingplatform.admin.dto.AdminProperty;
import com.housingplatform.admin.dto.AuditLogDto;
import com.housingplatform.admin.dto.HousingRequestDto;
import com.housingplatform.admin.dto.UpdateHousingRequestStatusRequest;
import com.housingplatform.admin.mapper.AdminMapper;
import com.housingplatform.auth.error.BadRequestException;
import com.housingplatform.auth.error.NotFoundException;
import com.housingplatform.auth.model.AuthenticatedUser;
import com.housingplatform.auth.service.AuthorizationService;
import com.housingplatform.hosts.dto.HostBooking;
import com.housingplatform.hosts.dto.HostDto;
import com.housingplatform.hosts.mapper.HostMapper;
import com.housingplatform.properties.dto.PropertyStatusChange;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AdminService {

  private final AdminRepositoryCustom adminRepository;
  private final AuthorizationService authorizationService;

  public AdminService(
      AdminRepositoryCustom adminRepository, AuthorizationService authorizationService) {
    this.adminRepository = adminRepository;
    this.authorizationService = authorizationService;
  }

  @Transactional(readOnly = true)
  public AdminDashboardStats getDashboardStats(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    Map<String, Integer> stats = adminRepository.getDashboardStats();
    return new AdminDashboardStats(
        stats.get("pendingProperties"),
        stats.get("pendingHosts"),
        stats.get("openBookings"),
        stats.get("openHousingRequests"));
  }

  @Transactional(readOnly = true)
  public List<AdminProperty> listProperties(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listProperties().stream().map(AdminMapper::toAdminProperty).toList();
  }

  @Transactional
  public PropertyStatusChange publishProperty(AuthenticatedUser user, UUID propertyId) {
    authorizationService.requireAdmin(user);
    var property = adminRepository.publishProperty(propertyId);
    if (property == null) {
      throw new BadRequestException("Property must be pending review before it can be published");
    }

    adminRepository.writeAuditLog(
        user.id(),
        "property.published",
        "property",
        property.getId(),
        Map.of("title", property.getTitle(), "slug", property.getSlug()));

    return AdminMapper.toPropertyStatusChange(property.getId(), property.getStatus());
  }

  @Transactional
  public PropertyStatusChange rejectPropertyReview(AuthenticatedUser user, UUID propertyId) {
    authorizationService.requireAdmin(user);
    var property = adminRepository.rejectPropertyReview(propertyId);
    if (property == null) {
      throw new BadRequestException("Property must be pending review before it can be rejected");
    }

    adminRepository.writeAuditLog(
        user.id(),
        "property.review_rejected",
        "property",
        property.getId(),
        Map.of("title", property.getTitle(), "slug", property.getSlug()));

    return AdminMapper.toPropertyStatusChange(property.getId(), property.getStatus());
  }

  @Transactional(readOnly = true)
  public List<AdminHost> listHosts(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listHosts().stream().map(AdminMapper::toAdminHost).toList();
  }

  @Transactional
  public HostDto approveHost(AuthenticatedUser user, UUID hostId) {
    authorizationService.requireAdmin(user);
    var host = adminRepository.approveHost(hostId);
    if (host == null) {
      throw new BadRequestException("Host must be pending before it can be approved");
    }

    adminRepository.writeAuditLog(
        user.id(),
        "host.approved",
        "host",
        host.getId(),
        Map.of("display_name", host.getDisplayName()));

    return HostMapper.toDto(host);
  }

  @Transactional(readOnly = true)
  public List<HostBooking> listBookings(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listBookings().stream()
        .map(
            row ->
                new HostBooking(
                    row.id(),
                    row.status(),
                    row.bookingType(),
                    row.checkIn(),
                    row.checkOut(),
                    row.guestCount(),
                    row.customerNotes(),
                    row.propertyTitle(),
                    row.roomName(),
                    row.totalKrw(),
                    row.createdAt()))
        .toList();
  }

  @Transactional(readOnly = true)
  public List<AdminPayment> listPayments(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listPayments().stream().map(AdminMapper::toAdminPayment).toList();
  }

  @Transactional(readOnly = true)
  public List<HousingRequestDto> listHousingRequests(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listHousingRequests().stream()
        .map(AdminMapper::toHousingRequest)
        .toList();
  }

  @Transactional
  public HousingRequestDto updateHousingRequestStatus(
      AuthenticatedUser user, UUID requestId, UpdateHousingRequestStatusRequest request) {
    authorizationService.requireAdmin(user);
    var updated = adminRepository.updateHousingRequestStatus(requestId, request.status());
    if (updated == null) {
      throw new NotFoundException("Housing request not found");
    }

    adminRepository.writeAuditLog(
        user.id(),
        "housing_request.status_updated",
        "housing_request",
        updated.getId(),
        Map.of("status", updated.getStatus().dbValue()));

    return AdminMapper.toHousingRequest(updated);
  }

  @Transactional(readOnly = true)
  public List<AuditLogDto> listAuditLogs(AuthenticatedUser user) {
    authorizationService.requireAdmin(user);
    return adminRepository.listAuditLogs().stream().map(AdminMapper::toAuditLog).toList();
  }
}

File: apps/backend-node/src/features/admin/admin.service.ts
import { Injectable } from '@nestjs/common';

import type { AuthUser } from '../../shared/auth/auth-user.model';
import { AuthorizationService } from '../../shared/auth/authorization.service';
import { BadRequestError, NotFoundError } from '../../shared/errors';
import type { HostBookingDto } from '../hosts/dto/host-booking.dto';
import type { HostDto } from '../hosts/dto/host.dto';
import { mapHost, mapHostBooking } from '../hosts/mappers/host.mapper';
import type { PropertyStatusChange } from '../properties/dto/property-status-change';
import type {
  AdminDashboardStatsDto,
  AdminHostDto,
  AdminPaymentDto,
  AdminPropertyDto,
  AuditLogDto,
  HousingRequestDto,
  UpdateHousingRequestStatusDto,
} from './dto';
import {
  mapAdminHost,
  mapAdminPayment,
  mapAdminProperty,
  mapAuditLog,
  mapHousingRequest,
  mapPropertyStatusChange,
} from './mappers/admin.mapper';
import { AdminRepository } from './admin.repository';

@Injectable()
export class AdminService {
  constructor(
    private readonly repository: AdminRepository,
    private readonly auth: AuthorizationService,
  ) {}

  async getDashboardStats(user: AuthUser): Promise<AdminDashboardStatsDto> {
    await this.auth.requireAdmin(user);
    return this.repository.getDashboardStats();
  }

  async listProperties(user: AuthUser): Promise<AdminPropertyDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listProperties();
    return rows.map(mapAdminProperty);
  }

  async publishProperty(
    user: AuthUser,
    propertyId: string,
  ): Promise<PropertyStatusChange> {
    await this.auth.requireAdmin(user);

    const propertyRow = await this.repository.publishProperty(propertyId);
    if (!propertyRow) {
      throw new BadRequestError(
        'Property must be pending review before it can be published',
      );
    }

    await this.repository.writeAuditLog({
      actorId: user.id,
      action: 'property.published',
      entityType: 'property',
      entityId: propertyRow.id,
      metadata: { title: propertyRow.title, slug: propertyRow.slug },
    });

    return mapPropertyStatusChange(propertyRow.id, propertyRow.status);
  }

  async rejectPropertyReview(
    user: AuthUser,
    propertyId: string,
  ): Promise<PropertyStatusChange> {
    await this.auth.requireAdmin(user);

    const propertyRow = await this.repository.rejectPropertyReview(propertyId);
    if (!propertyRow) {
      throw new BadRequestError(
        'Property must be pending review before it can be rejected',
      );
    }

    await this.repository.writeAuditLog({
      actorId: user.id,
      action: 'property.review_rejected',
      entityType: 'property',
      entityId: propertyRow.id,
      metadata: { title: propertyRow.title, slug: propertyRow.slug },
    });

    return mapPropertyStatusChange(propertyRow.id, propertyRow.status);
  }

  async listHosts(user: AuthUser): Promise<AdminHostDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listHosts();
    return rows.map(mapAdminHost);
  }

  async approveHost(user: AuthUser, hostId: string): Promise<HostDto> {
    await this.auth.requireAdmin(user);

    const host = await this.repository.approveHost(hostId);
    if (!host) {
      throw new BadRequestError('Host must be pending before it can be approved');
    }

    await this.repository.writeAuditLog({
      actorId: user.id,
      action: 'host.approved',
      entityType: 'host',
      entityId: host.id,
      metadata: { displayName: host.display_name },
    });

    return mapHost(host);
  }

  async listBookings(user: AuthUser): Promise<HostBookingDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listBookings();
    return rows.map(mapHostBooking);
  }

  async listPayments(user: AuthUser): Promise<AdminPaymentDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listPayments();
    return rows.map(mapAdminPayment);
  }

  async listHousingRequests(user: AuthUser): Promise<HousingRequestDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listHousingRequests();
    return rows.map(mapHousingRequest);
  }

  async updateHousingRequestStatus(
    user: AuthUser,
    requestId: string,
    request: UpdateHousingRequestStatusDto,
  ): Promise<HousingRequestDto> {
    await this.auth.requireAdmin(user);

    const updated = await this.repository.updateHousingRequestStatus(
      requestId,
      request.status,
    );
    if (!updated) {
      throw new NotFoundError('Housing request not found');
    }

    await this.repository.writeAuditLog({
      actorId: user.id,
      action: 'housing_request.status_updated',
      entityType: 'housing_request',
      entityId: updated.id,
      metadata: { status: updated.status },
    });

    return mapHousingRequest({
      id: updated.id,
      email: updated.email,
      desiredArea: updated.desired_area,
      checkIn: updated.check_in,
      checkOut: updated.check_out,
      budgetMax: updated.budget_max,
      accommodationType: updated.accommodation_type,
      notes: updated.notes,
      status: updated.status,
      createdAt: updated.created_at,
    });
  }

  async listAuditLogs(user: AuthUser): Promise<AuditLogDto[]> {
    await this.auth.requireAdmin(user);
    const rows = await this.repository.listAuditLogs();
    return rows.map(mapAuditLog);
  }
}
